### PR TITLE
Query: Foundational work on stateless query pipeline. The main change he...

### PR DESCRIPTION
**Pull request by anpete**
...re is to go from buffering value readers in the query buffer to having value readers move through a query as part of the result data flow.

QueryBuffer removal is pending #2016.

### CODE REVIEW

File: src/EntityFramework.Core/Query/QueryBuffer.cs [hunk 1]
```diff
@@ -135,7 +116,7 @@ public virtual void StartTracking(object entity)
             List<BufferedEntity> bufferedEntities;
             if (_byEntityInstance.TryGetValue(entity, out bufferedEntities))
             {
-                foreach (var bufferedEntity in bufferedEntities.Where(e => e != null))
+                foreach (var bufferedEntity in bufferedEntities.Skip(1).Where(e => e != null))
```
Comment (maumar) [inline src/EntityFramework.Core/Query/QueryBuffer.cs]: might be good to add comment here

Comment (anpete) [inline src/EntityFramework.Core/Query/QueryBuffer.cs]: This whole class is going away soon.

File: src/EntityFramework.Core/Query/ExpressionTreeVisitors/MemberAccessBindingExpressionTreeVisitor.cs [hunk 2]
```diff
@@ -37,16 +108,44 @@ protected override Expression VisitMemberExpression(MemberExpression memberExpre
...
                 var member = memberExpression.Member;
                 var typeInfo = newExpression.Type.GetTypeInfo();
 
-                if (typeInfo.IsGenericType
-                    && typeInfo.GetGenericTypeDefinition() == typeof(IAsyncGrouping<,>))
+                if (typeInfo.IsGenericType)
                 {
-                    member = typeInfo.GetDeclaredProperty("Key");
+                    var genericTypeDefinition = typeInfo.GetGenericTypeDefinition();
+                    var asyncGrouping = genericTypeDefinition == typeof(IAsyncGrouping<,>);
```
Comment (maumar) [inline src/EntityFramework.Core/Query/ExpressionTreeVisitors/MemberAccessBindingExpressionTreeVisitor.cs]: this could be inlined

**Comment by maumar** [comment]
:shipit:

**Pull request merged**